components/scream/src/share/grid/remap/horizontal_remap_utility.hpp

```diff
@@ -0,0 +1,564 @@
+  view_1d<Real>     weights;     // A view of the associated weights in the mapping.
+  Int               m_length;    // The size of this segment.
+
+/*---------------------------------------------*/
+  // Helper function to check a segment is valid.
+  bool check() {
+    EKAT_REQUIRE(source_dofs.extent(0)==m_length);
+    EKAT_REQUIRE(weights.extent(0)==m_length);
+    EKAT_REQUIRE(source_idx.extent(0)==m_length);
+    Real wgt = 0.0;
+    Kokkos::parallel_reduce("", m_length, KOKKOS_LAMBDA (const int& ii, Real& lsum) {
+      lsum += weights(ii);
+    },wgt);
+    // The sum of the weights should be = 1.0, note due to round-off error we accept to within a tolerance.
+    Real tol = std::numeric_limits<Real>::epsilon() * 100.0;
+    if (std::abs(wgt-1.0)>=tol) {
+      printf("ERROR: GSMAP: checking remap segment for DOF = %d, total weight = %e.\n",m_dof,wgt);
+    }
+    return std::abs(wgt-1.0)<tol;
+  }
+
+/*---------------------------------------------*/
+  template <typename ScalarT>
+  view_2d<ScalarT> apply_segment(
+      const view_3d<ScalarT>& source_data)
+  {
+    view_2d_host<ScalarT> ret("",source_data.extent(1),source_data.extent(2));
+    Kokkos::deep_copy(ret,0.0);
+    auto source_data_h = Kokkos::create_mirror_view(source_data);
+    auto source_idx_h  = Kokkos::create_mirror_view(source_idx);
+    auto weights_h     = Kokkos::create_mirror_view(weights);
+    Kokkos::deep_copy(source_data_h,source_data);
+    Kokkos::deep_copy(source_idx_h,source_idx);
+    Kokkos::deep_copy(weights_h   ,weights);
+    for (int ii=0; ii< m_length; ii++) {
+      int idx = source_idx_h(ii);
+      for (int nn=0; nn<source_data.extent(1); nn++) {
+        auto src_data_sub = ekat::subview(source_data_h,idx,nn);
+        for (int kk=0; kk<src_data_sub.extent(0); kk++) {
+          ret(nn,kk) = ret(nn,kk) + src_data_sub(kk)*weights_h(ii);
+        }
+      }
+    }
+    return ret;
+  }
+/*---------------------------------------------*/
+  template <typename ScalarT>
+  view_1d<ScalarT> apply_segment(
+      const view_2d<ScalarT>& source_data)
+  {
+    view_1d<ScalarT> ret("",source_data.extent(1));
+    auto ret_h = Kokkos::create_mirror_view(ret);
+    Kokkos::deep_copy(ret_h,0.0);
+    auto source_data_h = Kokkos::create_mirror_view(source_data);
+    auto source_idx_h  = Kokkos::create_mirror_view(source_idx);
+    auto weights_h     = Kokkos::create_mirror_view(weights);
+    Kokkos::deep_copy(source_data_h,source_data);
+    Kokkos::deep_copy(source_idx_h,source_idx);
+    Kokkos::deep_copy(weights_h   ,weights);
+    for (int ii=0; ii< m_length; ii++) {
+      int idx = source_idx_h(ii);
+      auto src_data_sub = ekat::subview(source_data_h,idx);
+      for (int kk=0; kk<src_data_sub.extent(0); kk++) {
+        ret_h(kk) = ret_h(kk) + src_data_sub(kk)*weights_h(ii);
+      }
+    }
+    Kokkos::deep_copy(ret,ret_h);
+    return ret;
+  }
+/*---------------------------------------------*/
+  template <typename ScalarT>
```

Comment:
Let's add a brief description of what this is doing. As simple as it is, it will help folks understand what's going on.